library/psa_crypto_random.h

```diff
@@ -0,0 +1,218 @@
+
+#elif defined(MBEDTLS_HMAC_DRBG_C)
+
+#include "mbedtls/hmac_drbg.h"
+#if defined(MBEDTLS_SHA512_C) && defined(MBEDTLS_SHA256_C)
+#include <limits.h>
+#if SIZE_MAX > 0xffffffff
+/* Looks like a 64-bit system, so prefer SHA-512. */
+#define MBEDTLS_PSA_HMAC_DRBG_MD_TYPE MBEDTLS_MD_SHA512
+#else
+/* Looks like a 32-bit system, so prefer SHA-256. */
+#define MBEDTLS_PSA_HMAC_DRBG_MD_TYPE MBEDTLS_MD_SHA256
+#endif
+#elif defined(MBEDTLS_SHA512_C)
+#define MBEDTLS_PSA_HMAC_DRBG_MD_TYPE MBEDTLS_MD_SHA512
+#elif defined(MBEDTLS_SHA256_C)
+#define MBEDTLS_PSA_HMAC_DRBG_MD_TYPE MBEDTLS_MD_SHA256
+#else
+#error "No hash algorithm available for HMAC_DBRG."
+#endif
+
+#else
+#error "No DRBG module available for the psa_crypto module."
+#endif
+
+#include "mbedtls/entropy.h"
+
+/** The type of the PSA DRBG context.
+ */
+#if defined(MBEDTLS_CTR_DRBG_C)
+typedef mbedtls_ctr_drbg_context mbedtls_psa_drbg_context_t;
+#elif defined(MBEDTLS_HMAC_DRBG_C)
+typedef mbedtls_hmac_drbg_context mbedtls_psa_drbg_context_t;
+#endif
+
+/** Initialize the PSA DRBG.
+ *
+ * \param p_rng        Pointer to the Mbed TLS DRBG state.
+ */
+static inline void mbedtls_psa_drbg_init( mbedtls_psa_drbg_context_t *p_rng )
+{
+#if defined(MBEDTLS_CTR_DRBG_C)
+    mbedtls_ctr_drbg_init( p_rng );
+#elif defined(MBEDTLS_HMAC_DRBG_C)
+    mbedtls_hmac_drbg_init( p_rng );
+#endif
+}
+
+/** Deinitialize the PSA DRBG.
+ *
+ * \param p_rng        Pointer to the Mbed TLS DRBG state.
+ */
+static inline void mbedtls_psa_drbg_free( mbedtls_psa_drbg_context_t *p_rng )
+{
+#if defined(MBEDTLS_CTR_DRBG_C)
+    mbedtls_ctr_drbg_free( p_rng );
+#elif defined(MBEDTLS_HMAC_DRBG_C)
+    mbedtls_hmac_drbg_free( p_rng );
+#endif
+}
+
+/** The type of the PSA random generator context.
+ *
+ * The random generator context is composed of an entropy context and
+ * a DRBG context.
+ */
+typedef struct
+{
+    void (* entropy_init )( mbedtls_entropy_context *ctx );
+    void (* entropy_free )( mbedtls_entropy_context *ctx );
+    mbedtls_entropy_context entropy;
+    mbedtls_psa_drbg_context_t drbg;
+} mbedtls_psa_random_context_t;
+
+/** Return random data.
+ *
+ * This function is suitable as the \p f_rng parameter to Mbed TLS functions
+ * that require a random generator. Use mbedtls_psa_random_state() to
+ * obtain the \p p_rng parameter.
+ *
+ * \param p_rng         The DRBG context. This must be
+ *                      mbedtls_psa_random_state( \c rng )
+ *                      where \c rng is a pointer to a
+ *                      ::mbedtls_psa_random_context_t structure.
+ * \param output        The buffer to fill.
+ * \param output_len    The length of the buffer in bytes.
+ *                      It must be at most #MBEDTLS_PSA_RANDOM_MAX_REQUEST.
+ *
+ * \retval              \c 0 on success.
+ * \return              \c MBEDTLS_ERR_xxx_DRBG_xxx or
+ *                      \c MBEDTLS_ERR_PLATFORM_xxx on failure.
+ */
+static inline int mbedtls_psa_get_random( void *p_rng,
```

Comment:
Otherwise, why not mbedtls_psa_get_random( mbedtls_psa_random_context_t *rng , ... ) ? Why exposing the DRBG context directly ?

Reply:
This is the signature of `f_rng` parameters to classic mbedtls functions.

Reply:
I didn't explain myself right. I understand the reason of `void *p_rng` but I am wondering why the value of the parameter is the address of a DRBG context and not the address of the `mbedtls_psa_random_context_t` context. With a `mbedtls_psa_random_context_t` context I think we can get rid of most if not all calls to `mbedtls_psa_random_state()`.

Reply:
`p_rng` has to be the DRBG context, otherwise `f_rng` could not be `mbedtls_xxx_drbg_random`, it would have to be a wrapper that extracts the DRBG context from the random context.

Reply:
Based on my previous comment, I don't think anything should change here.

Reply:
Actually neither my code nor your proposal will work for the follow-up where `mbedtls_psa_get_random` becomes usable as a random generator in classic code, since the definition of `mbedtls_psa_random_state` in the non-external case references a local variable inside `psa_crypto.c`. The pointer to the CTR or HMAC DRBG state needs to be accessible globally, either through a non-static function or through a global variable of its own.

Reply:
> Actually neither my code nor your proposal will work for the follow-up where `mbedtls_psa_get_random` becomes usable as a random generator in classic code, since the definition of `mbedtls_psa_random_state` in the non-external case references a local variable inside `psa_crypto.c`. 

Yes I have experienced that in PR 3933 and added an API to work around this (see [here](https://github.com/ARMmbed/mbedtls/pull/3933/commits/9c382b0ee022449db033811b43830390369492fc)). What do you mean by follow-up, the third PR in this series or after that?

> The pointer to the CTR or HMAC DRBG state needs to be accessible globally, either through a non-static function or through a global variable of its own.

A pointer to the `mbedtls_psa_random_context_t` is enough. I am probably missing something but I don't understand why we would keep exposing the CTR or HMAC DRBG states to the users of `mbedtls_psa_get_random`. To me, the state now is the mbedtls_psa_random_context_t structure.

Reply:
I replaced `mbedtls_psa_random_state` with a constant pointer, indirected through a macro. For the time being, these are internal interfaces, so the names and the exact nature of `mbedtls_psa_get_random` and `MBEDTLS_PSA_RANDOM_STATE` can change in step 3.